### PR TITLE
Implement parseInt and parseFloat

### LINKED ISSUE #827: Polyfills for parseFloat and parseInt
These are commonly used functions in existing TS codebases. It would be nice to have some polyfills for these to better support the transpilation of TS code without having to use lua specific function calls e.g. `tonumber`.

In addition to that they are just generally helpful functions.

### PR DESCRIPTION
**Pull request by Perryvw**
Closes #827

### CODE REVIEW

File: src/lualib/ParseInt.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    7: "^-?[0123456]*",
+    8: "^-?[01234567]*",
+    9: "^-?[012345678]*",
+    10: "^-?[0123456789]*",
+    11: "^-?[0123456789aA]*",
+    12: "^-?[0123456789aAbB]*",
+    13: "^-?[0123456789aAbBcC]*",
+    14: "^-?[0123456789aAbBcCdD]*",
+    15: "^-?[0123456789aAbBcCdDeE]*",
+    16: "^-?[0123456789aAbBcCdDeEfF]*",
```
Comment (ark120202) [inline src/lualib/ParseInt.ts:16]: Looks like it actually supports bases up to 36

File: src/lualib/ParseInt.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+};
+
+function __TS__ParseInt(this: void, numberString: string, base?: number): number {
+    const number = tonumber(string.match(numberString, __TS__parseInt_patterns[base ?? 10]), base);
```
Comment (ark120202) [inline src/lualib/ParseInt.ts:20]: When `base` is out of supported range (<2 or >36) it should return `undefined`

File: src/lualib/ParseFloat.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+function __TS__ParseFloat(this: void, numberString: string): number {
+    const number = tonumber(string.match(numberString, "^-?%d+%.?%d*"));
```
Comment (ark120202) [inline src/lualib/ParseFloat.ts:2]: JS implementation also handles `Infinity` and `-Infinity`

File: src/lualib/declarations/math.d.ts [hunk 4]
```diff
@@ -8,4 +8,7 @@ declare namespace math {
     function atan(y: number, x?: number): number;
 
     function atan2(y: number, x: number): number;
+
+    function ceil(x: number): number;
+    function floor(x: number): number;
```
Comment (ark120202) [inline src/lualib/declarations/math.d.ts:13]: Why not just `Math.*`?

File: test/unit/builtins/numbers.spec.ts [hunk 5]
```diff
@@ -78,3 +78,41 @@ test("number intersected method", () => {
 test("numbers overflowing the float limit become math.huge", () => {
     util.testExpression`1e309`.expectToMatchJsResult();
 });
+
+describe.each(["parseInt", "parseFloat"])("parse numbers with %s", parseFunction => {
+    const numberStrings = ["3", "3.0", "9", "42", "239810241", "-20391", "3.1415", "2.7182", "-34910.3"];
```
Comment (ark120202) [inline test/unit/builtins/numbers.spec.ts:83]: JS also seems to trim string first, so `" 1 "` is `1`

File: src/lualib/ParseInt.ts (hunk 2 above)
Comment (ark120202) [inline src/lualib/ParseInt.ts:20]: > `?? 10`

Citing [MDN page](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt):

> If radix is undefined, 0, or unspecified, JavaScript assumes the following:
> 
> 1. If the input string begins with "0x" or "0X" (a zero, followed by lowercase or uppercase X), radix is assumed to be 16 and the rest of the string is parsed as a hexidecimal number.
> 2. If the input string begins with "0" (a zero), radix is assumed to be 8 (octal) or 10 (decimal). Exactly which radix is chosen is implementation-dependent. ECMAScript 5 clarifies that 10 (decimal) should be used, but not all browsers support this yet. For this reason, always specify a radix when using parseInt.
> 3. If the input string begins with any other value, the radix is 10 (decimal).

File: src/lualib/declarations/math.d.ts (hunk 4 above)
Comment (Perryvw) [reply src/lualib/declarations/math.d.ts:13]: I know these methods do what we want, if we can have such extremely complicated functions like these as lualib functions I would rather not risk it and use the lowest level possible in lualib implementations.

**Review by ark120202 — APPROVED** [review]

**Pull request merged**